l10n_it_ddt/models/stock_picking_package_preparation.py

```diff
@@ -532,57 +528,60 @@ def _prepare_invoice_line(self, qty, invoice_id=None):
-                    self.product_id.categ_id.name
+        res = {}
+        if self.sale_line_id and self.sale_line_id.product_id and (
+            self.sale_line_id.product_id.property_account_income_id or
+            self.sale_line_id.product_id.categ_id.
+            property_account_income_categ_id
+        ):
+            # Without property_account_income_id or
+            # property_account_income_categ_id
+            # _prepare_invoice_line would fail
+            res = self.sale_line_id._prepare_invoice_line(qty)
+        else:
+            account = (
+                self.product_id.property_account_income_id or
+                self.product_id.categ_id.property_account_income_categ_id)
+            if not account:
+                if invoice_id:
+                    invoice = self.env['account.invoice'].browse(invoice_id)
+                    account = invoice.journal_id.default_credit_account_id
+            if not account:
+                raise UserError(
+                    _(
+                        'Please define income account for this product: "%s" '
+                        '(id:%d) - or for its category: "%s".'
+                    ) % (
+                        self.product_id.name, self.product_id.id,
+                        self.product_id.categ_id.name
+                    )
                 )
-            )
-
-        fpos = None
-        if self.sale_line_id:
-            fpos = (
-                self.sale_line_id.order_id.fiscal_position_id or
-                self.sale_line_id.order_id.partner_id.
-                property_account_position_id
-            )
-        if fpos:
-            account = fpos.map_account(account)
+            fpos = None
+            if self.sale_line_id:
+                fpos = (
+                    self.sale_line_id.order_id.fiscal_position_id or
+                    self.sale_line_id.order_id.partner_id.
+                    property_account_position_id
+                )
+            if fpos:
+                account = fpos.map_account(account)
+            res['account_id'] = account.id
 
-        res = {
+        res.update({
             'ddt_line_id': self.id,
             'name': self.name,
             'sequence': self.sequence,
             'origin': self.package_preparation_id.name or '',
-            'account_id': account.id,
             'price_unit': self.price_unit,
             'quantity': qty,
             'discount': self.discount,
             'uom_id': self.product_uom_id.id,
             'product_id': self.product_id.id or False,
             'invoice_line_tax_ids': [(6, 0, self.tax_ids.ids)],
-            'account_analytic_id': (
-                self.sale_line_id and
-                self.sale_line_id.order_id.project_id.id or False
-            ),
-            'analytic_tag_ids': [(
-                6, 0, self.sale_line_id and
-                self.sale_line_id.analytic_tag_ids.ids or []
-            )],
-        }
+        })
```

Comment:
non ho testato, ma se il res proviene dall'else, questi campi non resteranno poi vuoti?

Reply:
@sergiocorato se ```self.sale_line_id``` è vuoto, questi campi verrebbero comunque valorizzati con ```False``` e ```[]``` (e se ```self.sale_line_id``` è vuoto, il ```res``` proviene dall'else, dove questi 2 campi sarebbero comunque non valorizzati)

Potrebbe forse esserci il caso in cui ```self.sale_line_id.product_id.property_account_income_id``` e ```self.sale_line_id.product_id.categ_id.property_account_income_categ_id``` sono vuoti, mentre ```self.sale_line_id.order_id.project_id``` e/o ```self.sale_line_id.analytic_tag_ids``` sono valorizzati.
In questo caso, i campi verrebbero erroneamente lasciati vuoti.

Gestisco allora questo caso, grazie

Reply:
:+1: